app/queues/dns/dns-worker.server.ts

```diff
@@ -0,0 +1,144 @@
+import { UnrecoverableError, Worker } from 'bullmq';
+import { DnsRecordStatus } from '@prisma/client';
+
+import { redis } from '~/lib/redis.server';
+import logger from '~/lib/logger.server';
+import { deleteDnsRecordById, updateDnsRecordById } from '~/models/dns-record.server';
+import {
+  createDnsRecord,
+  upsertDnsRecord,
+  deleteDnsRecord,
+  getChangeStatus,
+} from '~/lib/dns.server';
+
+import type { DnsRecord } from '@prisma/client';
+
+type DnsOperation = 'create' | 'update' | 'delete';
+
+type DnsChangeData = {
+  dnsOperation: DnsOperation;
+  username: DnsRecord['username'];
+  type: DnsRecord['type'];
+  fqdn: string;
+  value: DnsRecord['value'];
+  changeId?: string;
+  dnsStatus?: string;
+};
+
+enum Step {
+  Route53Change,
+  WaitOnRoute53Change,
+  SyncChangeToDatabase,
+  Finished,
+}
+
+export type DnsJobData = DnsChangeData & { dnsRecordId: DnsRecord['id']; step?: Step };
+
+export const dnsWorker = new Worker<DnsJobData>(
+  'dns-queue',
+  async (job) => {
+    logger.debug('worker started');
+    let { step = Step.Route53Change } = job.data;
+
+    while (step !== Step.Finished) {
+      switch (step) {
+        case Step.Route53Change: {
+          const changeId = await executeChange({
+            ...job.data,
+          });
+          await job.update({ ...job.data, changeId, step: Step.WaitOnRoute53Change });
+        }
+
+        case Step.WaitOnRoute53Change: {
+          const dnsStatus = await waitOnChange(job.data.changeId);
+          await job.update({ ...job.data, dnsStatus, step: Step.SyncChangeToDatabase });
+        }
+
+        case Step.SyncChangeToDatabase: {
+          const { dnsRecordId, dnsOperation, dnsStatus } = job.data;
+          await syncChange(dnsRecordId, dnsOperation, dnsStatus);
+          step = Step.Finished;
+          return step;
+        }
+
+        default: {
+          throw new UnrecoverableError(`invalid step: ${step}`);
+        }
+      }
+    }
+  },
+  {
+    connection: redis,
+    limiter: {
+      max: 5,
+      duration: 1_000,
+    },
+  }
+);
+
+async function executeChange({ dnsOperation, username, type, fqdn, value }: DnsChangeData) {
+  fqdn = fqdn.toLowerCase();
```

Comment:
I initially lowercased the domain because of database, but removed this line in last commit because Route53 always converts domain to lowercase. I tested and confirmed.